### ISSUE TITLE
Wrapping Dexie methods in its promise implementation without getting setTimeout being triggered

### ISSUE THREAD

**Issue by AnimaLupi** [dfahlander/Dexie.js#91]
Hello again, this one is driving me crazy.

I'm wrapping some methods around the ones of your library in my application (which is based on the angular framework, so I took care of using your Promise implementation instead of $q until the very end of the chain, when transactions are already committed).

Problem is that I don't seem to be able to create a Promise of yours without having setTimeOut being called, breaking the nested transaction chain I'm trying to build with my wrappers.

I'll make a  (stupid) example:
```
this.getNodesToBeListed = function(ids) {
        var nodes = [];
        if (angular.isArray(ids) && (ids.length) > 0) {
// Transaction kinda irrelevant in this case, but let's pretend it's not
            return db.transaction('r', db.node, function () {
                // is it better to use .each instead of then?
                db.node.where(':id').anyOf(ids).toArray().then(function (res) {
                    if (angular.isArray(res)) {
                        angular.forEach(res, function (dbNode) {
                            var node = nodeService.buildFromDbData(dbNode);
                            if (node.isDraft()) {
                                nodes.push(node);
                            }
                        });
                    }
                });
                return nodes;
            });
        } else {
            // Trying to return a promise resolving to an empty array, but in fact I could return some more elaborated thing
            return Dexie.Promise.resolve(nodes);
        }
    };
```

Then I'd like to chain this method to another via `dbservice.getNodesToBeListed(ids).then(...)` and so on, all inside in a parent transaction.

The problem arise on the `Dexie.Promise.resolve(nodes)` which doesn't seem the correct way to do it.

I also tried a not so elegant:

```
new window.Dexie.Promise(function(resolve, reject) {
            resolve(nodes);
});
```

But it seems to still trigger the setTimeout.

Sorry if I'm being clueless here, the fact is that in angular I'm probably used to a different implementation where I would simply use a `return $q.when(nodes)` and save the day...

This problem of getting a resolved Promise without triggering setTimeOut and the subsequent indexedDB commit also applies to dynamic chains (in fact, that's the place where I discovered the issue, only to find it applied everywhere in my code).

(with dynamic chain I mean something that programmatically builds a sequential chain of promises, based on an array.reduce for example)

Is the thing I'm trying to do accomplisheable with your implementation of Promise?

I suspect it is and it's just me not figuring out the proper way...

Thanks a lot, I hope my question was clear enough and I didn't forget something...

EDIT: I'm assuming I'm doing it the wrong way as my transactions get aborted with the message "already closed transaction" and after adding some console.log in the Dexie source near the setTimeOut(s) (including asap) I see it triggered and I thought that was the issue... But maybe it's something that happens normally and it's in a place where it wouldn't distrupt the transaction, and I'm using correctly Dexie.Promise.resolve() to get a newly resolved promise?

**Comment by dfahlander** [comment]
Reason it calls setTimeout when doing executing a promise at 'root' level is to be guarantee being fully Promise A+ compliant. Internally this would generate issues with the indexedDB transactions because they can commit, so I use Dexie.Promise._rootExec() internally that will, at end call all callbacks instead of sending them to setTimeout().

To make sure setTimeout() is never called, make sure the root of your call stack is executed by
```js
Dexie.Promise.resolve().then(function(){
    // Do your calls here.
});
```
In your case, there are several ways to make sure to prohibit setTimeout() from being called. One is to use Dexie.Promise.resolve() at the start of the entire chain. Another way is to use Dexie.Promise._rootExec() at that same level. 

```js
// Code calling getNodesToBeListed():

Dexie.Promise.resolve().then(function (){
    return x.getNodesToBeListed (ids);
}).then(function (nodes) {
   // the nodes come here
});

```

Note however, that setTimeout() will be called before the initial then() is called, but after that no more within the chain. If this is a problem, you could use Dexie.Promise._rootExec() yourself and callbacks will fire directly before call is returned.

```js

// Code calling getNodesToBeListed():

Dexie.Promise._rootExec(function (){
    return x.getNodesToBeListed (ids).then(function(nodes) {
         // ...
    });
});


```

I am curious about what problems does setTimeout() cause you? I'm not so familiar with angular but has it to do with rendering stuff in angular?

**Comment by AnimaLupi** [comment]
Hello, thank for you answer and sorry for my late reply, got back to my PC just now to test what you described above.

First, about the Angular part, it's not causing issues with that, at least I think I managed to sort that out. Basically, Angular implementation of promises afaik DO use setTimeout (as it was also noted in another post here about using a generalized Promise implementation) to integrate it in its $digest service which takes care of DOM updating. I managed to work around that by using your Promise implementation and just wrap them in Angular ones at the end of the chain in a layer closer to the application Controller, manually calling the $apply function of Angular to update the DOM (which, as far as I know, it's more efficient than having it update on every step of the chain).

Problem is that I tried to use _rootExec but unfortunately that doesn't solve my issue.

Basically I have several methods that wrap around the DB calls.

Each of these methods normally uses a transaction to ensure atomicity of the operations included in itself.

I wrote all of these methods to return a promise with the result contained into it, to chain them togheter. The value in the promise is often the result of a calculation/transformation of the retrieved/inserted DB values (hence the need to wrap the static calculated/transformed result retrieved into a new promise with something like Dexie.Promise.resolve(values) or your proposed _rootExec).

When I chain the above methods, I'd like to wrap them all into a parent transaction. That way, I still get atomicity all over the single parts of the chain: while the single methods are already atomic (due the use of transaction in them), I want them to behave as a single transaction even when chained, so in case of failure, everything is rolled back.

I tried the _rootExec solution (as using .resolve at the start of the chain is not what I needed, I need it at the end of each single method I chain up to build the parent transaction) but I found out that its return value isn't a Promise (i.e. doesn't have a .then method I could use to chain the result into another method of the bigger chain), leading to same issue as before. Using _rootExec to return a freshly built Promise wouldn't ofc solve the problem as it would give same initial issue.

If I'm not completely misleaded (note: I'm not exluding I'm missing the point somewhere :-), there are a couple ways I could cope with this situation:

1) Rewrite each bigger, parent transaction to be a single method with all code contained in itself (kinda suck as it would lead to code duplication) to avoid the use of Promise.resolve(<static value>).

2) Give up on using transaction on the big parent promise and either live the rest of my life with finger crossed OR write some rollback code at application level, which could fail too and lead data in inconsistent state...

3) Extend the Promise class (maybe breaking A+ compatibility?) and create a Promise.staticResolve method which would return a statically resolved promise with supplied value without calling setTimeout. I took a quick glance at Bluebird promise implementation and I think I saw something along those lines.

Question is: do you think it's feasible and reasonable to implement point 3? Would that bring along some issues? Is there something special I need to take into account? I'm particularly worried about the PSD part which takes care of carrying transaction data.

Thank you again for any advice on the matter!

**Comment by dfahlander** [comment]
Wow you are really utilizing the cool parts :)

So if I understand you correct, you may do something like:

```js

db.transaction('rw', db.table, function() {
    return lowerLevelFn();
}).then(function (result) {
    updateViewWithResult(result);
});

function lowerLevelFn() {
    if (x) {
        return db.transaction('rw', db.table, function () {
            // do something in this sub transaction and return a result
        });
    } else {
        return Dexie.Promise.resolve(y); // return result directly
    }
}

```

If so, it is a bug if it doesnt work. Dexie.Promise.resolve() MUST, when being called from within a parent transaction, NOT do setTimeout() to resolve itself. Only if it was called directly.

I feel I really want to understand the issue and create a unit test that reproduce it and resolve the bug.

**Comment by AnimaLupi** [comment]
You wrote a cool lib, I'm doing my best to use the cool parts ;-)

That being said, I don't want to exclude I'm doing something wrong. I'm in the middle of a migration from a synchronized localStorage api to the indexedDb one with Dexie, lots of stuff to do so actually something could be messing things up if you say that setTimeout shouldn't be called in that context!

Give me a few mins and I'll try at least to set up a barebone JSfiddle to test the issue outside the rest of the app (I think I tried already, but better be safe than sorry), I'll let you know the results asap and in case post the link.

**Comment by dfahlander** [comment]
Tried to reproduce your scenario in a unit test. Failed though to reproduce it. [Unit test source](https://github.com/dfahlander/Dexie.js/blob/master/test/tests-transaction.js) (the last test in the file).

You can play it here: https://rawgit.com/dfahlander/Dexie.js/master/test/run-unit-tests.html?module=transaction&testNumber=104

**Comment by dfahlander** [comment]
If possible to trigger the issue, it would be great to have that unit test modified so that I have something to play with ;)

**Comment by AnimaLupi** [comment]
Hi David,

after JSfiddling for more than a couple of hours (so far for a "couple of mins", lol) maybe I found what my mistake was. (Fiddle at url http://jsfiddle.net/azt4gujv/3/ btw)

In the meantime you wrote the test and by checking it I got confirmation of my suspects.

I was trying to chain the Dexie.Promise.resolve while in a parent transaction in the broader sense of the term. That is, I was starting parent transaction in the parent method and then doing the chain of the single methods. Some of these (I suppose, have yet to check in my real app but I bet there lies the problem) gave back the promise WITHOUT wrapping it in a sub transaction, that is:

```js
function chainPart(obj) {
      if (obj.value == 'xyz') {
            return Dexie.Promise.resolve(obj);
      } else {
            return db.transaction('rw', db.table, function() {
                   return db.get(....);
            });
      }
}

function chainOfStuff(obj) {
     return db.transaction('rw', db.table, function() {
           return chainPart(obj).then(function(res) {
                 // Other chaining here....
           });
     });
}

```

BUT, if chainPart is written this way:

```js
function chainPart(obj) {
      return db.transaction('rw', db.table, function() {
            if (obj.value == 'xyz') {
                  return Dexie.Promise.resolve(obj);
            } else {
                  return db.get(....);
            }
      });
}
```

It does indeed work! I was wrongly making the assumption that the parent (chainOfStuff) transaction was working on the Dexie.Promise.resolve in chainPart without the need of the db.transaction wrapping there!

I finally got the hint by re-reading the phrase

`
Dexie.Promise.resolve() MUST, when being called from within a parent transaction, NOT do setTimeout() to resolve itself.
`

I was basically chaining a promise along the way without wrapping that single step into transaction!

Sorry for having taken your time on this issue when the problem was just with me missing the point of wrapping the single chain parts in transactions was required. Maybe it won't be a complete waste if this error of mine can be put in the docs as something that must NOT be done :-)

I'll check the code in my app and I'm sure I'll find the bug I introduced by this misinterpretation somewhere (lots of cheering when I'll do :-)

In the meantime, thank you once more for the help!
